cryptoki/src/session/object_management.rs

```diff
@@ -547,6 +548,179 @@ impl Session {
+    ///
+    /// * `object` - The [ObjectHandle] used to reference the object
+    /// * `attributes` - The list of attribute types to retrieve
+    ///
+    /// # Returns
+    ///
+    /// A vector of [Attribute] containing the values of the available attributes.
+    ///
+    /// # Note
+    ///
+    /// This method follows PKCS#11 spec: in the first call, it provides pre-allocated buffers
+    /// for attributes with a known fixed size, and NULL pointers for other attributes to query
+    /// their size.
+    ///
+    /// After the first call:
+    /// - Attributes that fit in pre-allocated buffers are ready (pValue != NULL, valid ulValueLen)
+    /// - Attributes with NULL pValue but valid ulValueLen need a second fetch
+    /// - Attributes with CK_UNAVAILABLE_INFORMATION are skipped
+    ///
+    /// In total, a maximum of 2 calls to C_GetAttributeValue are made.
+    ///
+    pub fn get_attributes(
+        &self,
+        object: ObjectHandle,
+        attributes: &[AttributeType],
+    ) -> Result<Vec<Attribute>> {
+        // Step 1: Build pass1 template
+        // - Pre-allocate buffers for attributes with known fixed size
+        // - Use NULL pointers for all other attributes to query their size
+        let mut buffers: Vec<Vec<u8>> = Vec::with_capacity(attributes.len());
+        let mut template1: Vec<CK_ATTRIBUTE> = Vec::with_capacity(attributes.len());
+
+        for attr_type in attributes.iter() {
+            if let Some(size) = attr_type.fixed_size() {
+                // We know the needed size, we allocate
+                let mut buffer = vec![0u8; size];
+                template1.push(CK_ATTRIBUTE {
+                    type_: (*attr_type).into(),
+                    pValue: buffer.as_mut_ptr() as *mut c_void,
+                    ulValueLen: size as CK_ULONG,
+                });
+                buffers.push(buffer);
+            } else {
+                // This is a variable size, we set length to 0 and set the buffer ptr to NULL
+                template1.push(CK_ATTRIBUTE {
+                    type_: (*attr_type).into(),
+                    pValue: std::ptr::null_mut(),
+                    ulValueLen: 0,
+                });
+                buffers.push(Vec::new());
+            }
+        }
+
+        // Step 2: Make pass1 call to C_GetAttributeValue
+        let rv1 = unsafe {
+            Rv::from(get_pkcs11!(self.client(), C_GetAttributeValue)(
+                self.handle(),
+                object.handle(),
+                template1.as_mut_ptr(),
+                template1.len().try_into()?,
+            ))
+        };
+
+        match rv1 {
+            Rv::Ok
+            | Rv::Error(RvError::BufferTooSmall)
+            | Rv::Error(RvError::AttributeSensitive)
+            | Rv::Error(RvError::AttributeTypeInvalid) => {
+                // acceptable - we'll inspect ulValueLen/pValue
+            }
+            _ => {
+                rv1.into_result(Function::GetAttributeValue)?;
```

Comment:
The `rv1` is not used anywhere after this block so I would rather keep using `rv` instead of `rv1` and `rv2` as it makes it easier to read and follow.